include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -337,4 +338,64 @@ def ImpliesOp : SMTOp<"implies", [Pure]> {
   let assemblyFormat = "$lhs `,` $rhs attr-dict";
 }
 
+class QuantifierOp<string mnemonic> : SMTOp<mnemonic, [
+  RecursivelySpeculatable,
+  RecursiveMemoryEffects,
+  SingleBlockImplicitTerminator<"smt::YieldOp">,
+]> {
+  let description = [{
+    This operation represents the }] # summary # [{ as described in the
+    [SMT-LIB 2.6 standard](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf).
+    It is part of the language itself rather than a theory or logic.
+
+    The operation specifies the names (as an attribute) and types (as the types
+    of the block arguments of the regions) of bound variables that may be
+    used in the 'body' of the operation. If a 'patterns' region is specified,
+    the block arguments must match the ones of the 'body' region and (other
+    than there) must be used at least once in the 'patterns' region. It may also
+    not contain any operations that bind variables, such as quantifiers. While
+    the 'body' region must always yield exactly one `!smt.bool`-typed value, the
+    'patterns' region can yield an arbitrary number (but at least one) of SMT
+    values.
+
+    The 'no_patterns' attribute is only allowed when no 'patterns' region is
+    specified and forbids the solver to generate and use patterns for this
+    quantifier.
+
+    The 'weight' attribute indicates the importance of this quantifier being
+    instantiated compared to other quantifiers that may be present. The default
+    value is zero.
+  }];
+
+  let arguments = (ins DefaultValuedAttr<I32Attr, "0">:$weight,
+                       UnitAttr:$noPattern,
+                       StrArrayAttr:$boundVarNames);
+  let regions = (region SizedRegion<1>:$body,
+                        MaxSizedRegion<1>:$patterns);
```

Comment:
My understanding is that `patterns` should be a variadic region, right?
For instance: 
![image](https://github.com/llvm/circt/assets/20801162/721912a6-2242-46d9-b621-66529a8a2b47)

Reply:
Makes sense. I didn't read this part carefully enough.
But isn't it required that in each (multi-)pattern all bound variables have to occur at least once? Why does the second `:pattern` only cover `x0`?

Reply:
Giving this example in the standard to Z3 yields a warning: `WARNING: (14,1): pattern does not contain all quantified variables.`
```
(declare-sort A 0)
(declare-fun r (A A) Bool)
(declare-fun p (A A) Bool)
(declare-const a A)

(assert (forall ((x0 A ) (x1 A ) (x2 A ))
  (! (=> (and (r x0 x1 ) (r x1 x2 )) (r x0 x2 ))
  :pattern ((r x0 x1 ) (r x1 x2 ))
  :pattern ((p x0 a ))
)))

(check-sat)
```

So, I guess, there is either a problem with the example in the standard or Z3 does not strictly adhere to the standard.
I think I'll drop the verifier that checks that all bound variables are used at least once to be consistent with the standard, and because it's not a hard error in Z3.